File: DesignDS/LRU.cpp
// Link: https://leetcode.com/problems/lru-cache/

class LRUCache {
public:
    int n;
    unordered_map<int, pair<list<int>::iterator, int> > mp;
    list<int> vec;
    LRUCache(int capacity) {
        n = capacity;
    }

    void update(int key){
        vec.erase(mp[key].first);
        vec.push_front(key);
        mp[key].first = vec.begin();
    }
    
    int get(int key) {
        if(mp.find(key) == mp.end()){
            return -1;
        }
        update(key);
        return mp[key].second;
    }
    
    void put(int key, int value) {
        if(mp.find(key) == mp.end()){
            vec.push_front(key);
            mp[key] = {vec.begin(), value};
            n--;
        }
        else{
            mp[key].second = value;
        }

        update(key);

        if(n < 0){
            int removeKey = vec.back();
            vec.pop_back();
            mp.erase(removeKey);
            n++;
        }
    }
};

/**
 * Your LRUCache object will be instantiated and called as such:
 * LRUCache* obj = new LRUCache(capacity);
 * int param_1 = obj->get(key);
 * obj->put(key,vecue);
 */

File: DesignDS/LRU.ts
// Using Js Map. Map maintains the order of insertion, so we can use it like a queue.
// Idea is to delete the key every time it is queried and add it again so it goes to the last.
// Then whenever we hit capacity, just remove the key from the top of the map because that key
// is for sure going to be the one that is LRU.
class LRUCache {
	private map = new Map<number, number>();
	constructor(private capacity: number) {}

	get(key: number): number {
		const value = this.map.get(key);
		if (value === undefined) {
			return -1;
		}

		this.map.delete(key);
		this.map.set(key, value);
		return value;
	}

	put(key: number, value: number): void {
		if (this.map.has(key)) {
			this.map.delete(key);
		}

		this.map.set(key, value);

		if (this.map.size > this.capacity) {
			const lruKey = this.map.keys().next().value;
			this.map.delete(lruKey);
		}
	}
}

/**
 * Your LRUCache object will be instantiated and called as such:
 * var obj = new LRUCache(capacity)
 * var param_1 = obj.get(key)
 * obj.put(key,value)
 */

/**
 * Your LRUCache object will be instantiated and called as such:
 * var obj = new LRUCache(capacity)
 * var param_1 = obj.get(key)
 * obj.put(key,value)
 */
